Content.Server/_DV/Augments/AugmentPowerCellSystem.cs

```diff
@@ -0,0 +1,114 @@
+    [Dependency] private readonly MobStateSystem _mobState = default!;
+    [Dependency] private readonly PowerCellSystem _powerCell = default!;
+    [Dependency] private readonly SharedBodySystem _body = default!;
+    [Dependency] private readonly SharedPopupSystem _popup = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+    }
+
+    public (Entity<AugmentPowerCellSlotComponent, OrganComponent, PowerCellSlotComponent> Organ, Entity<BatteryComponent>? Battery)? TryGetAugmentPowerCell(EntityUid body)
+    {
+        foreach (var organ in _body.GetBodyOrganEntityComps<AugmentPowerCellSlotComponent>(body))
+        {
+            if (!TryComp<PowerCellSlotComponent>(organ, out var powerCellSlot))
+                continue;
+
+            var entity = new Entity<AugmentPowerCellSlotComponent, OrganComponent, PowerCellSlotComponent>(organ.Owner, organ.Comp1, organ.Comp2, powerCellSlot);
+
+            if (_powerCell.TryGetBatteryFromSlot(organ, out var batteryUid, out var batteryComp))
+            {
+                return (entity, new(batteryUid.Value, batteryComp));
+            }
+            return (entity, null);
+        }
+        return null;
+    }
+
+    public (Entity<AugmentPowerCellSlotComponent, OrganComponent, PowerCellSlotComponent> Organ, Entity<BatteryComponent>? Battery)? TryGetAugmentPowerCellFromAugment(EntityUid augment)
+    {
+        if (!TryComp<OrganComponent>(augment, out var organ) || organ.Body is not {} uid)
+            return null;
+
+        return TryGetAugmentPowerCell(uid);
+    }
+
+    public bool TryDrawPower(EntityUid augment, float amount)
+    {
+        if (!TryComp<OrganComponent>(augment, out var organ) || organ.Body is not {} body)
+            return false;
+
+        if (TryGetAugmentPowerCellFromAugment(augment) is not (_, var battery))
+        {
+            _popup.PopupEntity(Loc.GetString("augments-no-power-cell-slot"), body, body);
+            return false;
+        }
+
+        if (battery is not {} insertedBattery)
+        {
+            _popup.PopupEntity(Loc.GetString("power-cell-no-battery"), body, body);
+            return false;
+        }
+
+        if (!_battery.TryUseCharge(insertedBattery.Owner, amount))
+        {
+            _popup.PopupEntity(Loc.GetString("power-cell-insufficient"), body, body);
+            return false;
+        }
+
+        return true;
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        var query = EntityQueryEnumerator<HasAugmentPowerCellSlotComponent>();
+        while (query.MoveNext(out var owner, out _))
+        {
+            if (_mobState.IsDead(owner))
+                continue;
+
+            var powerCell = TryGetAugmentPowerCell(owner);
+            if (powerCell is null) // tuples + nested nullable = hell for pattern matching so break it down
+                continue;
+            var augment = powerCell!.Value.Organ;
```

Comment:
`if (TryGetAugmentPowerCell(uid) is not (var organ, var battery)` should work but at the very least `is not {} tuple` and access tuple instead of powerCell!.Value

Reply:
you can't destructure tuples with `{} value`, and `not (var organ, var battery)` checks for `battery` being null, which is not desired here

Reply:
well do the second option then